CI: Check whether Python.h is included first in a file

DWesl opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->
This makes some allowance for including headers that include python as well

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
#20149 proposed this CI job, #20491 motivated actually getting this working.

#### What does this implement/fix?
<!--Please explain your changes.-->
There is a script, adapted from #20149, to check whether a given file includes system headers before Python.h.  This expands that to check headers before source files, and to check the source files for python constructs implying either an indirect path to Python.h or a new extension beyond numpy and pybind.  This then adds a CI job to perform that check on the contents of the SciPy repository.

#### Additional information
<!--Any additional information you think is important.-->
I made a brief attempt to flag duplicate includes, but stopped when I noticed there were reasons for doing that (one header includes `sys/time.h` on every platform but Windows, another includes `mutex` under two orthogonal conditions).

A few of the changes aren't necessary to include Python.h first, but do make it easier to automate the checking process.

A NumPy maintainer mentioned it is possible to do something similar with clang-format, but I don't know how to do that.

DWesl committed:
DEV: Add script to check that Python.h is included first

Based on script from #20149.

Doesn't know about Boost::Python yet.

DEV: Print name of file check_python_h_first is checking

DEV: Add numpy/ndarrayobject.h to list of Python-including headers

One file includes this first.

DWesl committed:
CI: Add CI job to check if Python.h is included first.

DWesl committed:
BLD: Move python-including headers earlier in their files.

BLD: Move python-including file earlier in io/_fast_matrix_market/src/_fmm_core.cpp

BLD: Move stdlib header after python-including headers in optimize/_pava/pava_pybind.cpp

BLD:BUG: Move python-including file to top of special/_special_ufuncs.cpp

DEV: Move python-including file earlier in wrap_g77_abi.c

Wasn't actually a problem, but this makes it easier for my script to check.

DEV: Move python including file earlier in wrap_dummy_g77_abi.c

Not actually a problem, but this makes it easier for my script to check.

STY: Remove duplicate headers at start of special/sf_error.cc

DWesl committed:
DEV: Extend the list of python-including headers as we go

Hopefully this reduces false-positives.

DWesl committed:
DEV: Add npy_ namespace to list of python-implying constructs

DWesl committed:
DEV: Tell script ni_support.h includes python

It does include python first, and this will note that.

DWesl committed:
DEV: Only warn once for python-implying constructs in a file

DWesl committed:
DEV: Don't warn for python constructs in headers

The file including the header should have already included Python.h (and whatever file for the specific constructs, or the compilation would fail everywhere)

DWesl committed:
DEV: Add very basic comment-parsing to python.h checker

DWesl committed:
CI: Check files for duplicate header includes.

This can cause errors if the header has no guard, or just slow things a bit if it does have a guard.

CI: Exclude boost_math and highs from header check.

Entirely too many false positives.

CI: Don't check fast_float

It seems to be a bunch of headers concatenated, which naturally results in many duplicate includes.

DWesl committed:
Revert "CI: Check files for duplicate header includes."

This reverts commit bd118fa45ce03555891ef517af65e9ccefc85720.

DWesl committed:
CI: Exclude HighsPy from checks.

DWesl committed:
STY: Fix style issues of tools/check_python_h_first.py.

j-bowhay reviewed (commented):
We could probably just tack this onto the lint CI job rather than an entirely new job?

DWesl committed:
STY: Fix linter errors

DWesl committed:
CI: Move Python.h checker into Lint job.

DWesl commented:
> We could probably just tack this onto the lint CI job rather than an entirely new job?

That sounds sensible

tylerjereddy commented on tools/check_python_h_first.py:
```diff
@@ -0,0 +1,114 @@
+                if "*/" in line:
+                    in_comment = False
+                continue
+            match = re.match(HEADER_PATTERN, line)
```
Just a drive by comment, but the code runs more than twice as fast for me locally on our codebase if I use the compiled regex directly:

```diff
--- a/tools/check_python_h_first.py
+++ b/tools/check_python_h_first.py
@@ -62,7 +62,7 @@ def check_python_h_included_first(name_to_check: str) -> int:
                 if "*/" in line:
                     in_comment = False
                 continue
-            match = re.match(HEADER_PATTERN, line)
+            match = HEADER_PATTERN.match(line)
             if match:
                 this_header = match.group(1)
                 if this_header in PYTHON_INCLUDING_HEADERS:
```

Apart from that, I didn't see any issues from playing with it for a few minutes and reading the diff.

DWesl replied:
```suggestion
            match =HEADER_PATTERN.match(line)
```
Was not aware I could do that.  Done.

DWesl committed:
PERF: Use regex method instead of module function

Not sure what `re.compile` is doing before the matching that takes as long as the regex

thalassemia commented:
The Accelerate failure is on me. Should be resolved in gh-20542.

DWesl commented:
Wait, that include order actually was important?  I could change the script to find and allow headers that included no other headers.

thalassemia commented:
It was important, but it should not have been haha. My PR should make it so that it no longer matters.

rgommers commented:
I couldn't reproduce the Accelerate job's build issue locally, so pushed a rebase on top of the now-merged gh-20542.

j-bowhay commented on .github/workflows/lint.yml:
```diff
@@ -47,3 +47,9 @@ jobs:
+    - name: Check that Python.h is first in any file including it.
+      shell: bash
+      run: |
+        python tools/check_python_h_first.py $(find scipy -name highspy -prune \
```
If the runtime is not too bad it could also be nice to add this to the `dev.py lint` task so the local tooling reflects what happens in CI

j-bowhay replied:
Sorry I should have been clearer, I meant here
https://github.com/scipy/scipy/blob/0b949514128c3790666862229d0f956009d63ed1/dev.py#L953-L957
not in the lint.py file itself

DWesl replied:
Okay, done now.

DWesl committed:
DEV: Add python.h checker to lint.py

DWesl committed:
DEV: Move check_python_h_first linter from lint.py to dev.py

DWesl committed:
DEV: Add --diff-against command-line flag like lint.py.

Also add cxx/hxx extensions, and handle the resulting new false positives.

DWesl committed:
STY:BUG: Move more python-including files earlier.

These got found by the new file types.

DWesl committed:
STY: Wrap lines in dev.py and add .[ch]xx files to CI

DWesl committed:
STY:FIX: Remove extra blank line added to lint.py

DWesl commented on tools/lint.py:
```diff
@@ -90,6 +90,23 @@ def run_cython_lint(files):
     return res.returncode, res.stdout
 
 
+def run_check_python_h_first(files):
+    # type: (list[str]) -> tuple[int, str]
+    """Run check_python_h_first on files."""
+    if not files:
+        return 0, ""
+    res = subprocess.run(
+        [
+            sys.executable,
+            os.path.join(os.path.dirname(__file__), "check_python_h_first.py")
+        ]
+        + list(files),
+        stdout=subprocess.PIPE,
+        encoding="utf-8"
+    )
+    return (res.returncode, res.stdout)
+
+
```
Given I deleted the code using this when the functionality got moved to dev.py, should I delete this as well?
```suggestion
```

rgommers replied:
yes, makes sense - committing. will push a few tweaks after.

rgommers committed:
Update tools/lint.py

[skip ci]

Co-authored-by: DWesl <22566757+DWesl@users.noreply.github.com>

rgommers committed:
DEV: fix running plain `python tools/check_python_h_first.py`

rgommers committed:
MAINT: undo now unnecessary change to tools/lint.py

rgommers committed:
MAINT: combine linter tasks in dev.py

[skip cirrus] [skip circle]

rgommers commented on scipy/io/_fast_matrix_market/src/_fmm_core.cpp:
```diff
@@ -2,7 +2,8 @@
 // Use of this source code is governed by the BSD 2-clause license found in the LICENSE.txt file.
 // SPDX-License-Identifier: BSD-2-Clause
 
-#include <Python.h>
```
This doesn't look correct, since there is direct usage of Python C API in this file. The linter also seems happy with the original version of the file. Unless I'm missing something here, I think you can revert this?

DWesl replied:
`_fmm_core.hpp` is included next, and includes `Python.h` first, which the checker script knows.  Either way would work.

DWesl replied:
I think the rationale for removing this where I kept the others was that I was the one to add this include, and recently.

rgommers replied:
Ah okay, fair enough. I think it's technically more correct to keep it, but both work - whatever you prefer.

rgommers commented on .github/workflows/lint.yml:
```diff
@@ -47,3 +47,11 @@ jobs:
+    - name: Check that Python.h is first in any file including it.
+      shell: bash
+      run: |
+        python tools/check_python_h_first.py $(find scipy \
+            \( -name highspy -o -name boost \) -prune \
+            -o -type f \( -iname \*.[ch] -o -name \*.[ch]pp -o -name \*.[ch]xx \
+                -o -name \*.cc -o -name \*.hh \) -print)
```
Can this be written without the `$(find ...)`? Easier to understand, and it then matches `python dev.py lint`. The exclusions don't seem needed for me, but if they are then maybe move them into the defaults of the script?

DWesl replied:
```suggestion
        python tools/check_python_h_first.py 
```
was how I tested the most recent changes locally; not *quite* the same as `python dev.py lint`, but the differences are primarily in whether the first include file must be `Python.h` or a file including `Python.h`

rgommers commented on scipy/spatial/ckdtree/src/count_neighbors.cxx:
```diff
@@ -1,3 +1,6 @@
+#include "ckdtree_decl.h"
```
These are all correct indeed - necessary because `"ckdtree_decl.h"` includes `npy_common.h`

rgommers commented on scipy/special/sf_error.cc:
```diff
@@ -1,8 +1,3 @@
-#include <Python.h>
-
-#include <stdarg.h>
-#include <stdlib.h>
```
This does not look correct. At least `stdarg` is needed for `va_start` & co. I don't immediately see if `stdlib` is used too.
Also @steppi why does this use `<stdarg.h>` rather than `<cstdarg>`?

DWesl replied:
The next four includes are these three and one more; I found this when attempting to check for duplicate includes.

rgommers replied:
Right, I just looked at the removals. Sorry for the noise - resolving.

rgommers reviewed (commented):
Thanks @DWesl. I pushed a few tweaks (please check) and added a few review comments on the changes to headers.

rgommers committed:
Update .github/workflows/lint.yml

[skip circle] [skip cirrus]

Co-authored-by: DWesl <22566757+DWesl@users.noreply.github.com>

rgommers reviewed (approved):
This is pretty clean now, and CI is happy. Let's give it a go! Thanks @DWesl, much appreciated
